src/protocol.rs

```diff
@@ -1,42 +1,541 @@
+    }
+
+    fn to_stream(&self, stream: &mut io::Write) -> io::Result<()> {
+        for segment in &self.0.segments() {
+            stream.write_u16::<BigEndian>(*segment)?;
+        }
+        Ok(())
+    }
+}
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct SCTPSegment(pub u16);
+
+derive_for_wrapping_segment!(SCTPSegment: Display, FromStr, From<u16>);
+impl AddressSegment for SCTPSegment {
+    const STREAM_LENGTH: usize = 2;
+
+    fn protocol(&self) -> Protocol {
+        Protocol::SCTP
+    }
+
+    fn from_stream(stream: &mut io::Read) -> Result<Self> {
+        Ok(SCTPSegment(stream.read_u16::<BigEndian>()?))
+    }
+
+    fn to_stream(&self, stream: &mut io::Write) -> io::Result<()> {
+        Ok(stream.write_u16::<BigEndian>(self.0)?)
+    }
+}
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct UDTSegment;
+
+derive_for_empty_segment!(UDTSegment: Display, FromStr, AddressSegment<UDT>);
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct UTPSegment;
+
+derive_for_empty_segment!(UTPSegment: Display, FromStr, AddressSegment<UTP>);
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct IPFSSegment(pub cid::Cid);
+
+derive_for_wrapping_segment!(IPFSSegment: Display, From<cid::Cid>);
+impl IPFSSegment {
+    fn _read_cid_polyfill(stream: &mut io::Read) -> Result<cid::Cid> {
+        // Read minimal CID length at start to check for CIDv0
+        let mut buf = [0u8; 34];
+        stream.read_exact(&mut buf)?;
+
+        if cid::Version::is_v0_binary(&buf) {
+            Ok(cid::Cid::from(&buf as &[u8])?)
+        } else {
+            // Do normal CID parsing using the already read data
+            let mut stream = Cursor::new(&buf as &[u8]).chain(stream);
+
+            // Read and parse CID header
+            let raw_version = stream.read_varint()?;
+            let raw_codec = stream.read_varint()?;
+            let _: u64 = stream.read_varint()?;
+
+            let version = cid::Version::from(raw_version)?;
+            let codec = cid::Codec::from(raw_codec)?;
+
+            let mh_len = stream.read_varint()?;
+
+            // Read CID hash data
+            // (Unsafe because data in `Vec` contains uninitialized memory
+            //  until we overwrite it with data read from `stream`.)
+            let mut hash = Vec::with_capacity(mh_len);
+            unsafe { hash.set_len(mh_len) };
+            stream.read_exact(hash.as_mut())?;
+
+            Ok(cid::Cid::new(codec, version, hash.as_slice()))
+        }
+    }
+}
+
+impl AddressSegment for IPFSSegment {
+    const STREAM_LENGTH: usize = 34;
+
+    fn protocol(&self) -> Protocol {
+        Protocol::IPFS
+    }
+
+    fn from_stream(stream: &mut io::Read) -> Result<Self> {
+        //FIXME: `cid::Prefix` creates its own `Cursor` instead of
+        //       overloading `io::Read`
+
+        // Read CID hash from stream
+        let cid = Self::_read_cid_polyfill(stream).map_err(|err| {
+            println!("CID Parsing failed!");
```

Comment:
Actually this is left-over debugging code 🙂 

Do you think it's worth adding logging for this anyhow?

Reply:
I don't know, I don't have a strong opinion about that. Usually I prefer to keep the logs more 'top-level', but it's up to you, cause you know better if it's needed or not

Reply:
Yes, now that I think about it again I'm going to remove it. The actual solution is fix up the `cid` crate so that we don't need `_read_cid_polyfill` anymore.